controllers/utils/multicloud_util.go

```diff
@@ -302,25 +318,74 @@ func DeleteMultiCloudObjects(
 	}
 
 	for _, mcv := range mcvList.Items {
-		multiCloudLog.Info("[DeleteMultiClusterObjects] Delete ManagedClusterView", "name", mcv.Name)
+		multiCloudLog.Info("[DeleteMultiClusterObjects] Delete ManagedClusterView", "ns", mcv.Namespace, "name", mcv.Name)
 		if err := c.Delete(ctx, &mcv); err != nil {
 			return err
 		}
 	}
-
 	return nil
 }
 
-// DeleteAllMultiCloudObjects cleans up views associated to all clusters in the list.
-func DeleteAllMultiCloudObjects(
-	ctx context.Context, c client.Client, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, allClustersForUpgrade []string) error {
-	// Only need to delete ManagedClusterView as ManagedClusterActions get deleted automatically.
-	for _, clusterName := range allClustersForUpgrade {
-		err := DeleteMultiCloudObjects(ctx, c, clusterGroupUpgrade, clusterName)
-		if err != nil {
+// DeleteManagedClusterActions cleans up actions associated to a cluster.
+func DeleteManagedClusterActions(
+	ctx context.Context, c client.Client, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, clusterName string) error {
+
+	var labels = map[string]string{
+		"openshift-cluster-group-upgrades/clusterGroupUpgrade": clusterGroupUpgrade.Namespace + "-" + clusterGroupUpgrade.Name}
+	opts := []client.ListOption{
+		client.InNamespace(clusterName),
+		client.MatchingLabels(labels),
+	}
+
+	mcaList := &actionv1beta1.ManagedClusterActionList{}
+	if err := c.List(ctx, mcaList, opts...); err != nil {
+		return err
+	}
+
+	for _, mca := range mcaList.Items {
+		multiCloudLog.Info("[DeleteMultiClusterObjects] Delete ManagedClusterAction", "ns", mca.Namespace, "name", mca.Name)
+		if err := c.Delete(ctx, &mca); err != nil {
 			return err
 		}
 	}
+	return nil
+}
+
+// FinalMultiCloudObjectCleanup cleans up views associated to all unsuccessful clusters from the current batch if known
+func FinalMultiCloudObjectCleanup(
+	ctx context.Context, c client.Client, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) error {
+	// Only need to cleanup for clusters that are not completed
+	succeededClusters := make(map[string]bool)
+	for _, clusterState := range clusterGroupUpgrade.Status.Clusters {
+		if clusterState.State == ClusterRemediationComplete {
+			succeededClusters[clusterState.Name] = true
+		}
+	}
+
+	// No current batch, clean up all just in case
+	if clusterGroupUpgrade.Status.Status.CurrentBatch == 0 {
```

Comment:
Hmm - how would this work when the CGU hasn't been enabled?

Reply:
It would be the same as before where we try cleaning up for every cluster all the time. I suppose we could add more conditional checks to further optimize the CGU delete case.

Reply:
I was thinking about the case where we're deleting the CGU without /before enabling it. Would that work or would it crash here because clusterGroupUpgrade.Status.Status is not defined?

Reply:
It wouldn't crash. If plan is populated, we clean up all. If not, nothing to do. Anyway, it's all changed now. Please review the updated logic.